### PR TITLE
Customers that have unsubscribed to the newsletter in the time span from when the queue was created and the actual sending of the newsletter still receive newsletters

### LINKED ISSUE #32116: Customers that have unsubscribed to the newsletter in the time span from when the queue was created and the actual sending of the newsletter still receive newsletters
### Preconditions (*)

1. Latest Magento install 2.4-develop

### Steps to reproduce (*)

1. Subscribe 2 customers to newsletter
2. Confirm their subscription
3. Create a newsletter queue that is to execute after a few minutes (say 5 for example)
4. Unsubscribe one of the customers (**before** the queue executes)

### Expected result (*)

1. Only the subscribed customer receives the newsletter email

### Actual result (*)

1. Both customers receive the newsletter email.

### Additional comment 
Due to harshness of GDPR laws and the sometimes enormous fees I'm marking this is S0 in my book.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by engcom-Charlie**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#32116

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/Newsletter/etc/di.xml [hunk 1]
```diff
@@ -32,4 +32,8 @@
     </type>
     <preference for="Magento\Newsletter\Model\SubscriptionManagerInterface"
                 type="Magento\Newsletter\Model\SubscriptionManager"/>
+    <type name="Magento\Newsletter\Model\Subscriber">
```
Comment (Den4ik) [inline app/code/Magento/Newsletter/etc/di.xml]: `Magento\Newsletter\Model\Subscriber` already defined. Please move plugin definition

File: app/code/Magento/Newsletter/Model/Plugin/RemoveSubscriberFromQueue.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public function afterUnsubscribe(Subscriber $subject, Subscriber $subscriber): Subscriber
+    {
+        if ($subscriber->dataHasChangedFor(self::STATUS)
+            && $subscriber->getSubscriberStatus() === Subscriber::STATUS_UNSUBSCRIBED
+        ) {
+            $this->connection->delete(
+                $this->connection->getTableName('newsletter_queue_link'),
+                ['subscriber_id = ?' => $subscriber->getId(), 'letter_sent_at IS NULL']
+            );
```
Comment (Den4ik) [inline app/code/Magento/Newsletter/Model/Plugin/RemoveSubscriberFromQueue.php]: Please, move this part into resource model that will allow reusing this code for third party modules

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi @engcom-Charlie. Thanks for PR. Can you have a look at my comment below?

Another suggestion for code improvement:
* Add call `Magento\Newsletter\Model\Subscriber->setStatusChanged(true)` into `Magento\Newsletter\Model\Subscriber->setStatus`
* Call `setStatus` instead of `setSubscriberStatus` at `unsubscribe` method
* Check `isStatusChanged` and compare with `Subscriber::STATUS_UNSUBSCRIBED` in the plugin

What do you think?

Thanks!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Integration Tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
Hi @engcom-Charlie
Thanks for your work and sorry that confuse you. I discussed changes with @sivaschenko

> it would be better to introduce a new class instead of adding new public method to existing API class. I believe in this case it can be easily done as the added method in quite independent

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/Newsletter/Model/RemoveSubscriberFromQueueLink.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * Removes subscriber from queue
+     *
+     * @param int $subscriberId
+     * @return void
+     */
+    public function execute(int $subscriberId): void
+    {
+        $this->connection->delete(
+            $this->connection->getTableName('newsletter_queue_link'),
```
Comment (sivaschenko) [inline app/code/Magento/Newsletter/Model/RemoveSubscriberFromQueueLink.php]: @engcom-Charlie I believe the table name should be retrieved from `ResourceConnection`

Comment (Den4ik) [reply app/code/Magento/Newsletter/Model/RemoveSubscriberFromQueueLink.php]: I think is better to add normal resource model `QueueLink` that extends `AbstractDb` and use `getMainTable` in query.
@sivaschenko what do you think?

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by Den4ik** [comment]
@magento run Functional Tests EE

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE

**Review by Den4ik — APPROVED** [review]
@engcom-Charlie Thanks for the work.

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-8815](https://jira.corp.magento.com/browse/ENGCOM-8815) has been created to process this Pull Request
:eight_spoked_asterisk: @Den4ik, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/Newsletter/Model/RemoveSubscriberFromQueueLink.php (hunk 3 above)
Comment (sivaschenko) [reply app/code/Magento/Newsletter/Model/RemoveSubscriberFromQueueLink.php]: I don't think we need to introduce a new class for a simple database operation.

Also, implementing granular commands is a better approach comparing to using resource models in terms of maintainability and extensibility.

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-8815](https://jira.corp.magento.com/browse/ENGCOM-8815) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Subscribe 2 customers to the newsletter;
2. Confirm their subscription;
3. Create a newsletter queue that is to execute after a few minutes (say 5 for example);
4. Unsubscribe one of the customers (before the queue executes);

**Before:** :heavy_multiplication_x: Both customers receive the newsletter email

Screenshot from 2021-02-25 15-14-07
Screenshot from 2021-02-25 15-14-15

**After:** :heavy_check_mark: An unsubscribed customer did not receive the email.
Only іsubscribed users have received emails.
Screenshot from 2021-02-26 09-29-23

**Pull request merged**